base/src/main/scala/clazz/Applicative.scala

```diff
@@ -2,8 +2,16 @@ package scato
 package clazz
 
 abstract class Applicative[F[_]] {
-  def apply: Apply[F]
+
   def pure[A](a: A): F[A]
+  def ap[A, B](fa: F[A])(f: F[A => B]): F[B]
+
+  def map[A, B](ma: F[A])(f: (A) => B): F[B] = ap(ma)(pure(f))
+
+  final val apply: Apply[F] = new Apply[F] {
```

Comment:
You are reaching the core principles of the encoding :-)

Your mostly right but this not achieved thru safe coerce, it's done thru the higher order natural transformation, for example `Applicative` ~~> `Apply`:
https://github.com/aloiscochard/scato/blob/master/base/src/main/scala/TCBase.scala#L31

We have to specify them, but the reason we have to it is for the same reason you asked your last question: we have to make an arbitrary decision about which instance to choose.

We know the instance are the same because we follow coherent design principles, so it does not really matter which one we choose, it's just that with have to define that bias for the implicit to resolve correctly.

In the same vain, that's why we should not use subtyping on the class them self, even if, as you said (and catched in the example!) it is actually possible.

Everything seems cool at first, but after a while you'll get the diamond problem in your face and you'll basically get back some issues from the previous encoding.

Hope that helps!